File: src/encoder.erl
-module(encoder).
-author("Piotr Ociepka").
-export([encode_exi/1]).
-on_load(init/0).
-compile(export_all).

init() ->
    Priv = code:priv_dir(exi),
    Nif = filename:join(Priv, "encoder_nif"),
    ok = erlang:load_nif(Nif, 0).

encode_exi([]) -> [];
encode_exi([{sd} | T]) ->
  encode_start_document(),
  encode_exi(T);
encode_exi([{se, Name} | T]) ->
  encode_start_element(Name),
  encode_exi(T);
encode_exi([{se, Name, Uri} | T]) ->
  encode_start_element_with_uri(Name, Uri),
  encode_exi(T);
encode_exi([{at, Name} | T]) ->
  encode_attribute(Name),
  encode_exi(T);
encode_exi([{at, Name, Uri} | T]) ->
  encode_attribute_with_uri(Name, Uri),
  encode_exi(T);
encode_exi([{sc} | T]) ->
  encode_self_contained(),
  encode_exi(T);
encode_exi([{ch, Value} | T]) ->
  encode_value(Value),
  encode_exi(T);
encode_exi([{ee} | T]) ->
  encode_end_element(),
  encode_exi(T);
encode_exi([{ed} | T]) ->
  encode_end_document(),
  encode_exi(T).


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% Initalizers
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

init_buffer() ->
    exit(nif_library_not_loaded).

init_stream() ->
  exit(nif_library_not_loaded).

serialize_header() ->
  exit(nif_library_not_loaded).

close_stream() ->
  exit(nif_library_not_loaded).

init_exi_header() ->
  exit(nif_library_not_loaded).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% Header options
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

set_cookie() ->
  exit(nif_library_not_loaded).

unset_cookie() ->
  exit(nif_library_not_loaded).

set_options() ->
  exit(nif_library_not_loaded).

unset_options() ->
  exit(nif_library_not_loaded).

set_version(_Version) ->
  exit(nif_library_not_loaded).

set_bit_packed_alignment() ->
  exit(nif_library_not_loaded).

set_byte_alignment() ->
  exit(nif_library_not_loaded).

set_precompression_alignment() ->
  exit(nif_library_not_loaded).

set_compression() ->
  exit(nif_library_not_loaded).

set_strict() ->
  exit(nif_library_not_loaded).

set_fragment() ->
  exit(nif_library_not_loaded).

set_preserved_comments() ->
  exit(nif_library_not_loaded).

set_preserved_pis() ->
  exit(nif_library_not_loaded).

set_preserved_dtd() ->
  exit(nif_library_not_loaded).

set_preserved_prefixes() ->
  exit(nif_library_not_loaded).

set_preserved_lexvalues() ->
  exit(nif_library_not_loaded).

set_schema_id() ->
  exit(nif_library_not_loaded).

set_block_size(_Size) ->
  exit(nif_library_not_loaded).

set_value_max_length(_Length) ->
  exit(nif_library_not_loaded).

set_partition_capacity(_Capacity) ->
  exit(nif_library_not_loaded).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% EXI events
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

encode_start_document() ->
  exit(nif_library_not_loaded).

encode_start_element(_Name) ->
  exit(nif_library_not_loaded).

encode_start_element_with_uri(_Name, _Uri) ->
  exit(nif_library_not_loaded).

encode_self_contained() ->
  exit(nif_library_not_loaded).

encode_attribute(_Name) ->
  exit(nif_library_not_loaded).

encode_attribute_with_uri(_Name, _Uri) ->
  exit(nif_library_not_loaded).

encode_value(_Value) ->
  exit(nif_library_not_loaded).

encode_end_element() ->
  exit(nif_library_not_loaded).

encode_end_document() ->
  exit(nif_library_not_loaded).

File: c_src/encoder.c
#include "EXISerializer.h"
#include "stringManipulate.h"
#include <stdio.h>
#include <string.h>

#include "encoder.h"

#define OUTPUT_BUFFER_SIZE 20000
EXIStream testStrm;
String uri;
String ln;
QName qname = {&uri, &ln, NULL};
char buf[OUTPUT_BUFFER_SIZE];
BinaryBuffer buffer;
EXITypeClass valueType;


void init_buffer() {
	buffer.buf = buf;
	buffer.bufLen = OUTPUT_BUFFER_SIZE;
	buffer.bufContent = 0;
	buffer.ioStrm.readWriteToStream = NULL;
	buffer.ioStrm.stream = NULL;
}

errorCode init_stream() {
	return serialize.initStream(&testStrm, buffer, NULL);
}

errorCode serialize_header() {
	return serialize.exiHeader(&testStrm);
}

errorCode close_stream() {
	return serialize.closeEXIStream(&testStrm);
}

/// HEADER ///

void init_exi_header() {
	serialize.initHeader(&testStrm);
}

void set_cookie() {
	testStrm.header.has_cookie = TRUE;
}

void unset_cookie() {
	testStrm.header.has_cookie = FALSE;
}

void set_options() {
	testStrm.header.has_options = TRUE;
}

void unset_options() {
	testStrm.header.has_options = FALSE;
}

void set_version(int version) {
	testStrm.header.version_number = version;
}

void set_bit_packed_alignment() {
	SET_ALIGNMENT(testStrm.header.opts.enumOpt, BIT_PACKED);
}

void set_byte_alignment() {
	SET_ALIGNMENT(testStrm.header.opts.enumOpt, BYTE_ALIGNMENT);
}

void set_precompression_alignment() {
	SET_ALIGNMENT(testStrm.header.opts.enumOpt, PRE_COMPRESSION);
}

void set_compression() {
	SET_COMPRESSION(testStrm.header.opts.enumOpt);
}

void set_strict() {
	SET_STRICT(testStrm.header.opts.enumOpt);
}

void set_fragment() {
	SET_FRAGMENT(testStrm.header.opts.enumOpt);
}

void set_preserved_comments() {
	SET_PRESERVED(testStrm.header.opts.preserve, PRESERVE_COMMENTS);
}

void set_preserved_pis() {
	SET_PRESERVED(testStrm.header.opts.preserve, PRESERVE_PIS);
}

void set_preserved_dtd() {
	SET_PRESERVED(testStrm.header.opts.preserve, PRESERVE_DTD);
}

void set_preserved_prefixes() {
	SET_PRESERVED(testStrm.header.opts.preserve, PRESERVE_PREFIXES);
}

void set_preserved_lexvalues() {
	SET_PRESERVED(testStrm.header.opts.preserve, PRESERVE_LEXVALUES);
}

void set_self_contained() {
	SET_SELF_CONTAINED(testStrm.header.opts.enumOpt);
}

void set_schema_id() {
	const String SCHEMA_ID_STR = {"http://www.ltu.se/exip", 22};
	testStrm.header.opts.schemaID = SCHEMA_ID_STR;
}

void set_block_size(int block_size) {
	testStrm.header.opts.blockSize = block_size;
}

void set_value_max_length(int value_max_length) {
	testStrm.header.opts.valueMaxLength = value_max_length;
}

void set_partition_capacity(int partition_capacity) {
	testStrm.header.opts.valuePartitionCapacity = partition_capacity;
}

/// EVENTS ///
errorCode encode_start_document() {
	return serialize.startDocument(&testStrm);
}

errorCode encode_start_element(char* name) {
	String local_name;
	asciiToString(name, &local_name, &testStrm.memList, FALSE);
	String elem_uri;
	getEmptyString(&elem_uri);
	qname.uri = &elem_uri;
	qname.localName = &local_name;
	return serialize.startElement(&testStrm, qname, &valueType);
}

errorCode encode_start_element_with_uri(char* name, char* uri) {
	String local_name;
	asciiToString(name, &local_name, &testStrm.memList, FALSE);
	String elem_uri;
	asciiToString(uri, &elem_uri, &testStrm.memList, FALSE);
	qname.uri = &elem_uri;
	qname.localName = &local_name;
	return serialize.startElement(&testStrm, qname, &valueType);
}

errorCode encode_self_contained() {
	return serialize.selfContained(&testStrm);
}

errorCode encode_attribute(char* name) {
	String local_name;
	asciiToString(name, &local_name, &testStrm.memList, FALSE);
	String elem_uri;
	getEmptyString(&elem_uri);
	qname.uri = &elem_uri;
	qname.localName = &local_name;
	return serialize.attribute(&testStrm, qname, TRUE, &valueType);
}

errorCode encode_attribute_with_uri(char* name, char* uri) {
	String local_name;
	asciiToString(name, &local_name, &testStrm.memList, FALSE);
	String elem_uri;
	asciiToString(uri, &elem_uri, &testStrm.memList, FALSE);
	qname.uri = &elem_uri;
	qname.localName = &local_name;
	return serialize.attribute(&testStrm, qname, TRUE, &valueType);
}

errorCode encode_value(char* value) {
	String val;
	asciiToString(value, &val, &testStrm.memList, FALSE);
	return serialize.stringData(&testStrm, val);
}

errorCode encode_end_element() {
	return serialize.endElement(&testStrm);
}

errorCode encode_end_document() {
	return serialize.endDocument(&testStrm);
}
